auth/api/iam/api.go

```diff
@@ -204,6 +207,75 @@ func (r Wrapper) RetrieveAccessToken(_ context.Context, request RetrieveAccessTo
 	return RetrieveAccessToken200JSONResponse(token), nil
 }
 
+func (r Wrapper) CreateDPoPProof(ctx context.Context, request CreateDPoPProofRequestObject) (CreateDPoPProofResponseObject, error) {
+	// check method and url
+	if request.Body.Method == "" {
+		return nil, core.InvalidInputError("missing method")
+	}
+	if request.Body.Url == "" {
+		return nil, core.InvalidInputError("missing url")
+	}
+	// check access token status
+	if request.Body.Token == "" {
+		return nil, core.InvalidInputError("missing token")
+	}
+
+	// extract DID from request path
+	ownDID, err := r.toOwnedDID(ctx, request.Did)
+	if err != nil {
+		return nil, err
+	}
+	// create new DPoP header
+	httpRequest, err := http.NewRequest(request.Body.Method, request.Body.Url, nil)
+	if err != nil {
+		return nil, core.InvalidInputError(err.Error())
+	}
+	dpop, err := r.auth.IAMClient().DPoPProof(ctx, *ownDID, *httpRequest, request.Body.Token)
+	return CreateDPoPProof200JSONResponse{Dpop: dpop}, err
+}
+
+func (r Wrapper) ValidateDPoPProof(_ context.Context, request ValidateDPoPProofRequestObject) (ValidateDPoPProofResponseObject, error) {
+	dpopToken, err := dpop.Parse(request.Body.Dpop)
+	if err != nil {
+		log.Logger().WithError(err).Debug("ValidateDPoPProof: failed to parse DPoP header")
+		return ValidateDPoPProof200JSONResponse{}, nil
+	}
+	if ok, err := dpopToken.Match(request.Body.Thumbprint, request.Body.Method, request.Body.Url); !ok {
+		log.Logger().Debugf("ValidateDPoPProof: %s", err.Error())
+		return ValidateDPoPProof200JSONResponse{}, nil
+	}
+	// check if ath claim matches hash of access_token
+	ath, ok := dpopToken.Token.Get(dpop.ATHKey)
+	if !ok {
+		log.Logger().Debug("ValidateDPoPProof: missing ath claim")
+		return ValidateDPoPProof200JSONResponse{}, nil
+	}
+	hash := hash2.SHA256Sum([]byte(request.Body.Token))
+	if ath != base64.RawURLEncoding.EncodeToString(hash.Slice()) {
+		log.Logger().Debug("ValidateDPoPProof: ath/token claim mismatch")
+		return ValidateDPoPProof200JSONResponse{}, nil
+	}
+	// check if the jti is already used, if not add it to the store for the duration of the access token lifetime
```

Comment:
That requires the resource server to do header inspection and read a fresh nonce from a response. I got the feeling this would be a burden on the implementing parties and leaks security stuff into the information layer. For security it's not required since our access tokens are short lived.